From ed1c680589d72f5762c76e33f679b4a367bc841a Mon Sep 17 00:00:00 2001
From: Nishant Sharma <nishant.sharma@arm.com>
Date: Mon, 4 Dec 2023 11:44:17 +0000
Subject: [PATCH] neoverse-rd/pcie_setup: enable io virtualisation block
 programming

The IO virtualisation block is made up of the SystemMemory Management
Unit (SMMU), Tower NCI and Generic Interrupt Controller (GIC) IPs. It
provides address translation support for IO devices.

This driver enables the support to configure the block dynamically.
The driver temporarily maps the entire memory carveout to a block and
does PCIe buswalk. During PCIe buswalk, the pcie_enumeration module
checks and cputures the resource requirement of each of the PCIe end
point. The module then uses this data to calculate and allocate
carveout regions for ECAM, MMIOL and MMIOH regions and program/update
the same in NCI and CMN.

Signed-off-by: Nishant Sharma <nishant.sharma@arm.com>
Change-Id: I2e68127dff49ab07b8d14517c79e686065c02a93

Signed-off-by: Jon Mason <jon.mason@arm.com>
Upstream-Status: Pending
---
 .../module/pcie_setup/CMakeLists.txt          |  19 +
 .../module/pcie_setup/Module.cmake            |  10 +
 .../pcie_setup/include/mod_pcie_setup.h       | 131 ++++
 .../module/pcie_setup/src/mod_pcie_setup.c    | 610 ++++++++++++++++++
 4 files changed, 770 insertions(+)
 create mode 100644 product/neoverse-rd/module/pcie_setup/CMakeLists.txt
 create mode 100644 product/neoverse-rd/module/pcie_setup/Module.cmake
 create mode 100644 product/neoverse-rd/module/pcie_setup/include/mod_pcie_setup.h
 create mode 100644 product/neoverse-rd/module/pcie_setup/src/mod_pcie_setup.c

diff --git a/product/neoverse-rd/module/pcie_setup/CMakeLists.txt b/product/neoverse-rd/module/pcie_setup/CMakeLists.txt
new file mode 100644
index 000000000000..a284d205462d
--- /dev/null
+++ b/product/neoverse-rd/module/pcie_setup/CMakeLists.txt
@@ -0,0 +1,19 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2024, Arm Limited and Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+add_library(${SCP_MODULE_TARGET} SCP_MODULE)
+
+target_include_directories(${SCP_MODULE_TARGET}
+    PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}/include"
+    PUBLIC "${CMAKE_SOURCE_DIR}/interface/address_remapper/"
+    PUBLIC "${CMAKE_SOURCE_DIR}/interface/cmn"
+    PUBLIC "${CMAKE_SOURCE_DIR}/product/neoverse-rd/interface/io_block/")
+
+target_sources(${SCP_MODULE_TARGET}
+               PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}/src/mod_pcie_setup.c")
+
+target_link_libraries(${SCP_MODULE_TARGET}
+    PRIVATE module-system-info module-pcie_discovery)
diff --git a/product/neoverse-rd/module/pcie_setup/Module.cmake b/product/neoverse-rd/module/pcie_setup/Module.cmake
new file mode 100644
index 000000000000..c28f8e2b9e7a
--- /dev/null
+++ b/product/neoverse-rd/module/pcie_setup/Module.cmake
@@ -0,0 +1,10 @@
+#
+# Arm SCP/MCP Software
+# Copyright (c) 2024, Arm Limited and Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+set(SCP_MODULE "pcie-setup")
+
+set(SCP_MODULE_TARGET "module-pcie-setup")
diff --git a/product/neoverse-rd/module/pcie_setup/include/mod_pcie_setup.h b/product/neoverse-rd/module/pcie_setup/include/mod_pcie_setup.h
new file mode 100644
index 000000000000..c387c442fced
--- /dev/null
+++ b/product/neoverse-rd/module/pcie_setup/include/mod_pcie_setup.h
@@ -0,0 +1,131 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2024, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Description:
+ *     Module for configuring PCIe rootports and its memory map.
+ */
+
+#ifndef MOD_PCIE_SETUP_H
+#define MOD_PCIE_SETUP_H
+
+#include <fwk_id.h>
+#include <fwk_module_idx.h>
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/*!
+ * \brief Defines base address and size of carveout.
+ */
+
+struct mod_pcie_setup_carveout {
+    /*! Start of the carveout */
+    uint64_t start;
+    /*! Size of the carveout */
+    uint64_t size;
+};
+
+/*!
+ * \brief Defines regions for ECAM, MMIOL, MMIOH and Bus.
+ */
+struct mod_pcie_setup_mmap {
+    struct mod_pcie_setup_carveout ecam;
+    struct mod_pcie_setup_carveout mmiol;
+    struct mod_pcie_setup_carveout mmioh;
+    struct mod_pcie_setup_carveout bus;
+};
+
+/*!
+ * \brief configuration for a single end point and memory map for non io macro.
+ */
+struct mod_pcie_setup_ep_config {
+    /*! Valid flag to indicate if the mapping is valid to program */
+    bool valid;
+    /*! Allow non-secure access
+     *
+     * By default only secure accesses are allowed. Set this to true to allow
+     * non-secure access as well.
+     */
+    bool allow_ns_access;
+    /*! ID of the Root port node. */
+    uint8_t rp_node_id;
+};
+
+/*!
+ * \brief Configuration information used for programming IO virtualisation
+ *        block.
+ */
+struct mod_pcie_setup_config {
+    /*! NCI GVP base */
+    uint64_t reg_base;
+    /*! ID of the node in the CMN mesh configuration */
+    unsigned int cmn_node_id;
+    /*!* ID of the source node in NCI */
+    uint8_t nci_source_node_id;
+    /*!* ID of the Io Virtualization block. */
+    uint8_t block_id;
+    /*! End point config for RPs */
+    struct mod_pcie_setup_ep_config *ep_config;
+};
+
+/*!
+ * \brief Platform notification source and notification id
+ *
+ * \details On platforms that require platform configuration to access the SDS
+ *          memory regions, the platform notification can be subscribed. This is
+ *          optional for a platform and if provided as module configuration
+ *          data, the configuration will be done only after this notification is
+ *          processed.
+ */
+struct mod_pcie_setup_platform_notification {
+    /*! Identifier of the notification id */
+    const fwk_id_t notification_id;
+
+    /*! Identifier of the module sending the notification */
+    const fwk_id_t source_id;
+};
+
+/*!
+ * \brief Module configuration data used for per chip resource allocation.
+ */
+struct mod_pcie_setup_resource_info {
+    /*! ID of concrete implementation of CMN interface. */
+    fwk_id_t cmn_id;
+    /*! ID of concrete implementation of IO Block interface. */
+    fwk_id_t io_block_id;
+    /*! ID of concrete implementation of Address Remapper. */
+    fwk_id_t address_remapper_id;
+    /*! ID of CMN map_io_region api. */
+    fwk_id_t cmn_api_id;
+    /*! ID of IO Block io_block_setup api. */
+    fwk_id_t io_block_api_id;
+    /*! ID of IO Block address remapper rw api. */
+    fwk_id_t address_remapper_api_id;
+    /*! Offset added to the address. */
+    uint64_t translation;
+    /*! Bank for allocating carveouts for ECAM, MMIOL and MMIOH. */
+    struct mod_pcie_setup_mmap mmap;
+    /*! Number of endpoints in each IO Block. */
+    uint8_t ep_count;
+    /*! Platform notification source and notification id */
+    struct mod_pcie_setup_platform_notification plat_notification;
+};
+
+/*!
+ * \brief Module configuration.
+ */
+struct mod_pcie_setup_module_config {
+    /*! Number of resource information configuration. One per chip. */
+    size_t resource_info_count;
+    /*!
+     * Resource configuration common for all the IO virtualsation blocks in a
+     * chip.
+     */
+    struct mod_pcie_setup_resource_info *resource_info;
+};
+
+#endif /* MOD_PCIE_SETUP_H */
diff --git a/product/neoverse-rd/module/pcie_setup/src/mod_pcie_setup.c b/product/neoverse-rd/module/pcie_setup/src/mod_pcie_setup.c
new file mode 100644
index 000000000000..e722119a0546
--- /dev/null
+++ b/product/neoverse-rd/module/pcie_setup/src/mod_pcie_setup.c
@@ -0,0 +1,610 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2024, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Description:
+ *     Neoverse-RD PCIe Setup.
+ */
+
+#include <mod_pcie_discovery.h>
+#include <mod_pcie_setup.h>
+#include <mod_system_info.h>
+
+#include <interface_address_remapper.h>
+#include <interface_cmn.h>
+#include <interface_io_block.h>
+
+#include <fwk_assert.h>
+#include <fwk_event.h>
+#include <fwk_id.h>
+#include <fwk_log.h>
+#include <fwk_macros.h>
+#include <fwk_mm.h>
+#include <fwk_module.h>
+#include <fwk_module_idx.h>
+#include <fwk_notification.h>
+#include <fwk_status.h>
+
+#include <fmw_cmsis.h>
+
+#include <inttypes.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#define MOD_NAME "[PCIE_SETUP] "
+
+struct mod_pcie_setup_context {
+    /* Pointer to the element configuration pointers (table) */
+    struct mod_pcie_setup_config **config;
+
+    /* APIs required to do PCIe bus walk and resource discovery */
+    struct mod_pcie_discovery_api *pcie_discovery_api;
+
+    /* Pointer to the module config. */
+    const void *module_config;
+
+    /* Pointer to the chip specific module info. */
+    struct mod_pcie_setup_resource_info resource_info;
+
+    /* Pointer to system info APIs */
+    struct mod_system_info_get_info_api *system_info_api;
+
+    /* Pointer to system info configuration. */
+    const struct mod_system_info *system_info;
+
+    /* API to program PCIe memory region in GVP NCI. */
+    struct interface_io_block_memmap_api *io_block_memmap_api;
+
+    /* APIs to configure CMN moudle. */
+    struct interface_cmn_memmap_rnsam_api *cmn_memmap_rnsam_api;
+
+    /* AP remap Address read write APIs */
+    struct interface_address_remapper_rw_api *remapper_rw_api;
+};
+
+struct mod_pcie_setup_context pcie_setup_context;
+
+static void allocate_ecam_address(
+    struct mod_pcie_setup_mmap *mmap,
+    struct pcie_mmap_size *mmap_size)
+{
+    struct mod_pcie_setup_resource_info *res_info;
+
+    res_info = &pcie_setup_context.resource_info;
+
+    if (mmap_size->ecam != 0) {
+        if (mmap_size->ecam > res_info->mmap.ecam.size) {
+            FWK_LOG_ERR(MOD_NAME "No memory left to allocate for ECAM");
+            fwk_unexpected();
+        }
+        mmap->ecam.start = res_info->mmap.ecam.start;
+        mmap->ecam.size = mmap_size->ecam;
+        res_info->mmap.ecam.start += mmap_size->ecam;
+        res_info->mmap.ecam.size -= mmap_size->ecam;
+    } else {
+        /*
+         * Reset the values to zero to prevent caller from using garbage values.
+         */
+        mmap->ecam.start = 0;
+        mmap->ecam.size = 0;
+    }
+}
+
+static void allocate_mmiol_address(
+    struct mod_pcie_setup_mmap *mmap,
+    struct pcie_mmap_size *mmap_size)
+{
+    struct mod_pcie_setup_resource_info *res_info;
+
+    res_info = &pcie_setup_context.resource_info;
+
+    if (mmap_size->mmiol != 0) {
+        if (mmap_size->mmiol > res_info->mmap.mmiol.size) {
+            FWK_LOG_ERR(MOD_NAME "No memory left to allocate for MMIOL");
+            fwk_unexpected();
+        }
+        mmap->mmiol.start = res_info->mmap.mmiol.start;
+        mmap->mmiol.size = mmap_size->mmiol;
+        res_info->mmap.mmiol.start += mmap_size->mmiol;
+        res_info->mmap.mmiol.size -= mmap_size->mmiol;
+    } else {
+        /*
+         * Reset the values to zero to prevent caller from using garbage values.
+         */
+        mmap->mmiol.start = 0;
+        mmap->mmiol.size = 0;
+    }
+}
+
+static void allocate_mmioh_address(
+    struct mod_pcie_setup_mmap *mmap,
+    struct pcie_mmap_size *mmap_size)
+{
+    struct mod_pcie_setup_resource_info *res_info;
+
+    res_info = &pcie_setup_context.resource_info;
+
+    if (mmap_size->mmioh != 0) {
+        if (mmap_size->mmioh > res_info->mmap.mmioh.size) {
+            FWK_LOG_ERR(MOD_NAME "No memory left to allocate for MMIOH");
+            fwk_unexpected();
+        }
+        mmap->mmioh.start = res_info->mmap.mmioh.start;
+        mmap->mmioh.size = mmap_size->mmioh;
+        res_info->mmap.mmioh.start += mmap_size->mmioh;
+        res_info->mmap.mmioh.size -= mmap_size->mmioh;
+    } else {
+        /*
+         * Reset the values to zero to prevent caller from using garbage values.
+         */
+        mmap->mmioh.start = 0;
+        mmap->mmioh.size = 0;
+    }
+}
+
+static void allocate_bus(
+    struct mod_pcie_setup_mmap *mmap,
+    struct pcie_mmap_size *mmap_size)
+{
+    struct mod_pcie_setup_resource_info *res_info;
+
+    res_info = &pcie_setup_context.resource_info;
+
+    if (mmap_size->bus != 0) {
+        if (mmap_size->bus > res_info->mmap.bus.size) {
+            FWK_LOG_ERR(MOD_NAME "No bus left to allocate");
+            fwk_unexpected();
+        }
+        mmap->bus.start = res_info->mmap.bus.start;
+        mmap->bus.size = mmap_size->bus;
+        res_info->mmap.bus.start += mmap_size->bus;
+        res_info->mmap.bus.size -= mmap_size->bus;
+    } else {
+        /*
+         * Reset the values to zero to prevent caller from using garbage values.
+         */
+        mmap->bus.start = 0;
+        mmap->bus.size = 0;
+    }
+}
+
+/*
+ * This takes in the size required for each of the region and then allocates
+ * regions for them from platform provided configuration.
+ */
+static inline void get_address_range(
+    struct mod_pcie_setup_mmap *mmap,
+    struct pcie_mmap_size *mmap_size)
+{
+    allocate_ecam_address(mmap, mmap_size);
+    allocate_mmiol_address(mmap, mmap_size);
+    allocate_mmioh_address(mmap, mmap_size);
+    allocate_bus(mmap, mmap_size);
+}
+
+static int map_region_in_io_block(
+    enum interface_io_block_carveout_type carveout_type,
+    struct mod_pcie_setup_config *config,
+    uint64_t address,
+    uint64_t size,
+    uint8_t index,
+    uint8_t *region)
+{
+    struct interface_io_block_carveout_info carveout_info = {
+        .base = address,
+        .size = size,
+        .carveout_type = carveout_type,
+        .target_id = config->ep_config[index].rp_node_id
+    };
+    struct interface_io_block_setup_mmap mmap_info = {
+        .io_block_address = config->reg_base,
+        .carveout_info = &carveout_info,
+        .region_count = 1,
+        .source_id = config->nci_source_node_id,
+    };
+    int status;
+
+    status = pcie_setup_context.io_block_memmap_api->map_region(&mmap_info);
+    if (status == FWK_SUCCESS) {
+        *region = carveout_info.region_id;
+    }
+
+    return status;
+}
+
+static int unmap_region_in_io_block(
+    struct mod_pcie_setup_config *config,
+    uint8_t region)
+{
+    struct interface_io_block_carveout_info carveout_info = {
+        .region_id = region,
+    };
+    struct interface_io_block_setup_mmap mmap_info = {
+        .io_block_address = config->reg_base,
+        .carveout_info = &carveout_info,
+        .region_count = 1,
+        .source_id = config->nci_source_node_id,
+    };
+    int status;
+
+    status = pcie_setup_context.io_block_memmap_api->unmap_region(&mmap_info);
+
+    return status;
+}
+
+static int discover_and_configure_pcie_device(
+    struct mod_pcie_setup_config *config,
+    uint8_t index,
+    struct mod_pcie_setup_mmap *mmap)
+{
+    struct interface_address_remapper_rw_api *remapper_rw_api;
+    struct mod_pcie_discovery_api *pcie_discovery_api;
+    struct mod_pcie_setup_resource_info *res_info;
+    struct pcie_discovery_rw_api pcie_rw_api;
+    struct pcie_mmap_size mmap_size = { 0 };
+    uint8_t region;
+    int status;
+
+    res_info = &pcie_setup_context.resource_info;
+
+    status = map_region_in_io_block(
+        INTERFACE_IO_BLOCK_CARVEOUT_TYPE_ECAM,
+        config,
+        res_info->mmap.ecam.start,
+        res_info->mmap.ecam.size,
+        index,
+        &region);
+    if (status != FWK_SUCCESS) {
+        FWK_LOG_ERR(MOD_NAME "NCI mapping failed");
+        return status;
+    }
+
+    remapper_rw_api = pcie_setup_context.remapper_rw_api;
+    pcie_rw_api.read8 = remapper_rw_api->read8;
+    pcie_rw_api.write8 = remapper_rw_api->write8;
+    pcie_rw_api.read32 = remapper_rw_api->read32;
+    pcie_rw_api.write32 = remapper_rw_api->write32;
+
+    pcie_discovery_api = pcie_setup_context.pcie_discovery_api;
+    pcie_discovery_api->calculate_resource(
+        res_info->mmap.ecam.start + res_info->translation,
+        &mmap_size,
+        res_info->mmap.bus.start,
+        &pcie_rw_api);
+    get_address_range(mmap, &mmap_size);
+
+    status = unmap_region_in_io_block(config, region);
+    if (status != FWK_SUCCESS) {
+        return status;
+    }
+
+    if (mmap->ecam.size != 0) {
+        status = map_region_in_io_block(
+            INTERFACE_IO_BLOCK_CARVEOUT_TYPE_ECAM,
+            config,
+            mmap->ecam.start,
+            mmap->ecam.size,
+            index,
+            &region);
+        if (status != FWK_SUCCESS) {
+            FWK_LOG_ERR(MOD_NAME "ECAM Mapping in NCI failed");
+            return status;
+        }
+    }
+
+    if (mmap->mmiol.size != 0) {
+        status = map_region_in_io_block(
+            INTERFACE_IO_BLOCK_CARVEOUT_TYPE_MMIOL,
+            config,
+            mmap->mmiol.start,
+            mmap->mmiol.size,
+            index,
+            &region);
+        if (status != FWK_SUCCESS) {
+            FWK_LOG_ERR(MOD_NAME "MMIOL Mapping in NCI failed");
+            return status;
+        }
+    }
+
+    if (mmap->mmioh.size != 0) {
+        status = map_region_in_io_block(
+            INTERFACE_IO_BLOCK_CARVEOUT_TYPE_MMIOH,
+            config,
+            mmap->mmioh.start,
+            mmap->mmioh.size,
+            index,
+            &region);
+        if (status != FWK_SUCCESS) {
+            FWK_LOG_ERR(MOD_NAME "MMIOH Mapping in NCI failed");
+            return status;
+        }
+    }
+
+    return FWK_SUCCESS;
+}
+
+static int configure_pcie_endpoint(struct mod_pcie_setup_config *config)
+{
+    struct interface_cmn_memmap_rnsam_api *cmn_memmap_rnsam_api;
+    struct mod_pcie_setup_mmap ep_mmap;
+    struct mod_pcie_setup_resource_info *res_info;
+    uint64_t ecam_start;
+    uint64_t mmiol_start;
+    uint64_t mmioh_start;
+    size_t ecam_size;
+    size_t mmiol_size;
+    size_t mmioh_size;
+    uint8_t index;
+    int status;
+
+    res_info = &pcie_setup_context.resource_info;
+
+    ecam_start = res_info->mmap.ecam.start;
+    mmiol_start = res_info->mmap.mmiol.start;
+    mmioh_start = res_info->mmap.mmioh.start;
+    ecam_size = 0;
+    mmiol_size = 0;
+    mmioh_size = 0;
+
+    cmn_memmap_rnsam_api = pcie_setup_context.cmn_memmap_rnsam_api;
+    status = cmn_memmap_rnsam_api->map_io_region(
+        res_info->mmap.ecam.start,
+        res_info->mmap.ecam.size,
+        config->cmn_node_id);
+    if (status != FWK_SUCCESS) {
+        return status;
+    }
+
+    for (index = 0; index < res_info->ep_count; index++) {
+        if (config->ep_config[index].valid) {
+            memset(&ep_mmap, 0, sizeof(ep_mmap));
+            status =
+                discover_and_configure_pcie_device(config, index, &ep_mmap);
+            if (status != FWK_SUCCESS) {
+                return status;
+            }
+            ecam_size += ep_mmap.ecam.size;
+            mmiol_size += ep_mmap.mmiol.size;
+            mmioh_size += ep_mmap.mmioh.size;
+        }
+    }
+
+    if (ecam_size != 0) {
+        status = cmn_memmap_rnsam_api->map_io_region(
+            ecam_start, ecam_size, config->cmn_node_id);
+        if (status != FWK_SUCCESS) {
+            return status;
+        }
+    }
+
+    if (mmiol_size != 0) {
+        status = cmn_memmap_rnsam_api->map_io_region(
+            mmiol_start, mmiol_size, config->cmn_node_id);
+        if (status != FWK_SUCCESS) {
+            return status;
+        }
+    }
+
+    if (mmioh_size != 0) {
+        status = cmn_memmap_rnsam_api->map_io_region(
+            mmioh_start, mmioh_size, config->cmn_node_id);
+        if (status != FWK_SUCCESS) {
+            return status;
+        }
+    }
+
+    return FWK_SUCCESS;
+}
+
+static int configure_pcie_ecam_mmio_space(
+    struct mod_pcie_setup_config *config,
+    unsigned int element)
+{
+    int status = FWK_SUCCESS;
+
+    FWK_LOG_INFO(MOD_NAME "Configuring PCIe: %d", element);
+    status = configure_pcie_endpoint(config);
+
+    return status;
+}
+
+/* Framework APIs. */
+static int mod_pcie_setup_init(
+    fwk_id_t module_id,
+    unsigned int block_count,
+    const void *data)
+{
+    const struct mod_pcie_setup_module_config *module_config;
+
+    if (block_count == 0) {
+        /* There must be at least one pcie setup config data */
+        fwk_unexpected();
+        return FWK_E_PARAM;
+    }
+
+    if (data == NULL) {
+        fwk_unexpected();
+        return FWK_E_DATA;
+    }
+
+    pcie_setup_context.module_config = data;
+    module_config = data;
+    if ((module_config == NULL) || (module_config->resource_info_count == 0)) {
+        return FWK_E_PARAM;
+    }
+
+    /*
+     * Copy first instance of the resource info for binding. target API and
+     * module ids are common for all the chip variants.
+     */
+    memcpy(
+        &pcie_setup_context.resource_info,
+        module_config->resource_info,
+        sizeof(pcie_setup_context.resource_info));
+    pcie_setup_context.config =
+        fwk_mm_calloc(block_count, sizeof(struct mod_pcie_setup_config *));
+    if (pcie_setup_context.config == NULL) {
+        return FWK_E_NOMEM;
+    }
+
+    return FWK_SUCCESS;
+}
+
+static int mod_pcie_setup_element_init(
+    fwk_id_t element_id,
+    unsigned int unused,
+    const void *data)
+{
+    struct mod_pcie_setup_config *config;
+
+    config = (struct mod_pcie_setup_config *)data;
+    if ((config == NULL) || (config->reg_base == 0)) {
+        fwk_unexpected();
+        return FWK_E_DATA;
+    }
+
+    pcie_setup_context.config[fwk_id_get_element_idx(element_id)] = config;
+
+    return FWK_SUCCESS;
+}
+
+static int mod_pcie_setup_bind(fwk_id_t id, unsigned int round)
+{
+    struct mod_pcie_setup_resource_info *res_info;
+    int status;
+
+    res_info = &pcie_setup_context.resource_info;
+
+    status = fwk_module_bind(
+        res_info->address_remapper_id,
+        res_info->address_remapper_api_id,
+        &pcie_setup_context.remapper_rw_api);
+    if (status != FWK_SUCCESS) {
+        return status;
+    }
+
+    status = fwk_module_bind(
+        res_info->cmn_id,
+        res_info->cmn_api_id,
+        &pcie_setup_context.cmn_memmap_rnsam_api);
+    if (status != FWK_SUCCESS) {
+        return status;
+    }
+
+    status = fwk_module_bind(
+        FWK_ID_MODULE(FWK_MODULE_IDX_SYSTEM_INFO),
+        FWK_ID_API(FWK_MODULE_IDX_SYSTEM_INFO, MOD_SYSTEM_INFO_GET_API_IDX),
+        &pcie_setup_context.system_info_api);
+    if (status != FWK_SUCCESS) {
+        return status;
+    }
+
+    status = fwk_module_bind(
+        res_info->io_block_id,
+        res_info->io_block_api_id,
+        &pcie_setup_context.io_block_memmap_api);
+    if (status != FWK_SUCCESS) {
+        return status;
+    }
+
+    return fwk_module_bind(
+        FWK_ID_MODULE(FWK_MODULE_IDX_PCIE_DISCOVERY),
+        FWK_ID_API(
+            FWK_MODULE_IDX_PCIE_DISCOVERY,
+            MOD_PCIE_DISCOVERY_API_IDX_CALCULATE_RESOURCES),
+        &pcie_setup_context.pcie_discovery_api);
+}
+
+static int mod_pcie_setup_start(fwk_id_t id)
+{
+    const struct mod_pcie_setup_module_config *module_config;
+    struct mod_pcie_setup_resource_info *res_info;
+    struct mod_pcie_setup_config *config;
+    unsigned int element;
+    unsigned int chip_id;
+    int status;
+
+    if (fwk_id_is_type(id, FWK_ID_TYPE_MODULE)) {
+        status = pcie_setup_context.system_info_api->get_system_info(
+            &pcie_setup_context.system_info);
+        if (status != FWK_SUCCESS) {
+            return status;
+        }
+
+        chip_id = pcie_setup_context.system_info->chip_id;
+        module_config = pcie_setup_context.module_config;
+        if (chip_id >= module_config->resource_info_count) {
+            FWK_LOG_ERR(MOD_NAME "Module configuration not valid.");
+            return FWK_E_PARAM;
+        }
+
+        memcpy(
+            &pcie_setup_context.resource_info,
+            &module_config->resource_info[chip_id],
+            sizeof(pcie_setup_context.resource_info));
+
+        return FWK_SUCCESS;
+    }
+
+    fwk_assert(fwk_module_is_valid_element_id(id));
+
+    element = fwk_id_get_element_idx(id);
+
+    res_info = &pcie_setup_context.resource_info;
+
+    if (fwk_id_type_is_valid(res_info->plat_notification.source_id)) {
+        /* Bind to the platform notification that is needed to start module
+         * configuration. */
+        status = fwk_notification_subscribe(
+            res_info->plat_notification.notification_id,
+            res_info->plat_notification.source_id,
+            id);
+    } else {
+        config = pcie_setup_context.config[element];
+        status = configure_pcie_ecam_mmio_space(config, element);
+    }
+
+    return status;
+}
+
+static int mod_pcie_setup_process_notification(
+    const struct fwk_event *event,
+    struct fwk_event *resp_event)
+{
+    struct mod_pcie_setup_resource_info *res_info;
+    struct mod_pcie_setup_config *config;
+    unsigned int element;
+    int status;
+
+    fwk_assert(!fwk_id_is_type(event->target_id, FWK_ID_TYPE_MODULE));
+
+    res_info = &pcie_setup_context.resource_info;
+    if (fwk_id_is_equal(
+            event->id, res_info->plat_notification.notification_id)) {
+        status = fwk_notification_unsubscribe(
+            event->id, event->source_id, event->target_id);
+        if (status != FWK_SUCCESS) {
+            return status;
+        }
+
+        element = fwk_id_get_element_idx(event->target_id);
+        config = pcie_setup_context.config[element];
+        status = configure_pcie_ecam_mmio_space(config, element);
+        if (status != FWK_SUCCESS) {
+            return status;
+        }
+    }
+
+    return FWK_SUCCESS;
+}
+
+const struct fwk_module module_pcie_setup = {
+    .type = FWK_MODULE_TYPE_DRIVER,
+    .init = mod_pcie_setup_init,
+    .element_init = mod_pcie_setup_element_init,
+    .bind = mod_pcie_setup_bind,
+    .start = mod_pcie_setup_start,
+    .process_notification = mod_pcie_setup_process_notification,
+};
